File: Medium/48. Rotate Image.java
class Solution {
    public void rotate(int[][] matrix) {
        int size = matrix.length;
        for (int r = 0; r < size / 2; r++) {
            int curSize = size - r * 2;
            for (int c = r; c < r + curSize - 1; c++) {
                int top = matrix[r][c]; // (1, 1)
                int right = matrix[c][size - 1 - r]; // (1, 2)
                int bottom = matrix[size - 1 - r][size - 1 - c]; // (2, 2)
                int left = matrix[size - 1 - c][r]; // (2, 1)
                matrix[r][c] = left;
                matrix[c][size - 1 - r] = top;
                matrix[size - 1 - r][size - 1 - c] = right;
                matrix[size - 1 - c][r] = bottom;
            }
        }
    }
}

File: Medium/48. Rotate Image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        for r in range(n // 2):
            for c in range(r, n - 1 - r):
                t = matrix[r][c]
                matrix[r][c] = matrix[n - 1 - c][r]
                matrix[n - 1 - c][r] = matrix[n - 1 - r][n - 1 - c]
                matrix[n - 1 - r][n - 1 - c] = matrix[c][n - 1 - r]
                matrix[c][n - 1 - r] = t
